nvflare/fuel/f3/cellnet/core_cell.py

```diff
@@ -868,6 +920,55 @@ def register_request_cb(self, channel: str, topic: str, cb, *args, **kwargs):
             raise ValueError(f"specified request_cb {type(cb)} is not callable")
         self.req_reg.set(channel, topic, Callback(cb, args, kwargs))
 
+    def encrypt_payload(self, message: Message):
+
+        if not message.get_header(MessageHeaderKey.SECURE, False):
+            return
+
+        payload_len = message.get_header(MessageHeaderKey.PAYLOAD_LEN)
+        if payload_len is not None:
+            # Prevent double encryption
+            return
+
+        dest = message.get_header(MessageHeaderKey.DESTINATION)
+        if not dest:
+            raise RuntimeError("Message destination missing")
+
+        target = FQCN.get_root(dest)
+
+        if message.payload is None:
+            message.payload = bytes(0)
+
+        payload_len = len(message.payload)
+        message.add_headers({MessageHeaderKey.PAYLOAD_LEN: payload_len})
+
+        target_cert = self.cert_ex.get_certificate(target)
+        message.payload = self.secure_message.encrypt(target_cert, message.payload)
+        self.logger.debug(f"Payload ({payload_len} bytes) is encrypted ({len(message.payload)} bytes)")
+
+    def decrypt_payload(self, message: Message):
+
+        if not message.get_header(MessageHeaderKey.SECURE, False):
+            return
+
+        payload_len = message.get_header(MessageHeaderKey.PAYLOAD_LEN)
+        if payload_len is None:
+            # Message is already decrypted
+            return
+
+        message.remove_header(MessageHeaderKey.PAYLOAD_LEN)
+
+        origin = message.get_header(MessageHeaderKey.ORIGIN)
+        if not origin:
+            raise RuntimeError("Message origin missing")
+
+        origin = FQCN.get_root(origin)
```

Comment:
Perhaps you should directly pass origin to "cert_exchanger" so it's up to it how to do it - using root or not.

Reply:
Fixed.